text/0079-cdk-2.0.md

```diff
@@ -0,0 +1,337 @@
+4. Generate module `package.json` - Create the package.json for the build of `monocdk-experiment`. This has no dependencies and includes `constructs` and `@types/node` as devDependencies.
+5. Install dependencies - Run `npm install`.
+6. Compile with jsii
+7. Stage for local consumption - copies the compiled `.js` files from the build directory back to the repository in the `packages/monocdk-experiment/staging` directory. This allows local modules that depend on `monocdk-experiment` to build against it.
+8. Package - module is packaged as a normal jsii package using `jsii-pacmak` in the repositories packaging step using `pack.sh`.
+
+All `@aws-cdk/` namespaced libraries will continue to be published as normal.
+
+### 2. Pre-release
+
+`monocdk-experiment` is renamed to `aws-cdk-lib`. The version of this module now becomes v2.0 and is tagged as a "pre-release" version. This has varying formats for different language targets.
+
+#### NodeJS
+version number: `2.0.0-beta.1`
+deploying: `npm publish --tag beta`
+installing: `npm install aws-cdk-lib@beta`
+package.json:
+```sh
+"dependencies": {
+  "aws-cdk-lib": "~2.0.0-beta.1"
+}
+```
+**Note: `~` ranges match all versions with the same major, minor, and patch numbers with a greater prerelease number. For example, the above would match `2.0.0-beta.2`. but not `2.0.1-beta.2`.**
+
+#### Java
+version number: `2.0.0-beta-1`
+pom.xml:
+```xml
+<dependencies>
+    <dependency>
+        <groupId>software.amazon.awscdk</groupId>
+        <artifactId>lib</artifactId>
+        <version>2.0.0-beta-1</version>
+    </dependency>
+</dependencies>
+```
+
+#### Dotnet
+version number: `2.0.0-beta.1`
+.csproj:
+```xml
+  <ItemGroup>
+    <PackageReference Include="Amazon.CDK.Lib" Version="2.0.0-beta.1" />
+  </ItemGroup>
+```
+
+#### Python
+version numbeer: `2.0b1`
+requirements.txt: 
+```
+aws-cdk.lib==2.0b1
+```
+
+All `@aws-cdk/` namespaced libraries will continue to be published with current `1.x.x` versions.
+
+### Stable
+
+After a period of testing and gathering feedback from users, `aws-cdk-lib` will be promoted to a stable release with the version `2.0.0`. All other modules, excluding those under the `@aws-cdk/` namespace, will also get promoted to v2.0, though will remain backwards compatible with `@aws-cdk/` modules.
+
+All `@aws-cdk/` namespaced libraries will continue to be published with current `1.x.x` versions.
+
+### v1 Deprecation
+
+`@aws-cdk/` namespaced modules will continue to receive non-breaking changes and be released regularly.
+
+When v1 module support is discontinued will ultimately be dictated by the introduction of some feature or change that the team agrees is worth the cost of dropping support and requires us to do so. Alternatively, if the burden of maintaining v1 module support is no longer warranted by the number of users continuing to rely on them.
```

Comment:
I'm not convinced we're doing the right thing here.

> dictated by the introduction of some feature or change that the team agrees is worth the cost of dropping support and requires us to do so

I'd rather us plan out when this happens, rather than pivot this on some feature/change that may or may not happen in the future.

> if the burden of maintaining v1 module support is no longer warranted by the number of users continuing to rely on them

The way I'm reading this is - "not in any reasonable timeline". We've this happen with releases other in other tools or ecosystem (e.g., it took a long time python2.7 could be deprecated).

We should a take a different approach instead - define and announce a sunset period and/or a maintenance period for 1.0 and stop releases after that. 
We should also consider an upgrade tool for customers to move from 1.0 to 2.0 (similar to the monocdk rewrite?) in all our supported languages.

Another benefit we will get from this is that deprecated code and packages can actually be deleted from our code base, and we'll no longer have to maintain backwards compat.

Reply:
I tend to agree and also becoming a bit concerned about our ability to actually hide deprecated APIs from v2.0 without really modifying the code. As part of #7708 and #7736 I ran into a few situations where stable APIs extend an experimental API (e.g. `assets.IAsset`). Apart from the fact that I think we should probably prohibit this, it will be impossible to hide `IAsset` even though it's marked `deprecated`. We have quite a few places like that in our code and I don't know how this will actually work.

An additional case is the constructs compatiblity layer, which will also require some actual code changes to get rid of.

I am leaning more and more towards a period of transitioning from a v1.0 code base to a v2.0 code base in which we can actually do some clean ups and leave the old stuff behind some how. This, naturally, will require us to plan better for this transition and be explicit about what we intend to backport. It also means that we may need to continue to release v1.x for a while from a branch.

Reply:
I'm leaning towards agreeing with both of you. I think we should agree on a period of continuing support for v1 modules and announce that up front. Additionally the cost of hiding deprecated APIs during build time appears to become more complex the more we look into it.

We could implement whatever we can with the v2 modules building from the v1 modules. Then we can branch the two, without creating a git branch, by copying all of the v1 modules src into the `aws-cdk-lib` modules src directory and change the way we are building to just build from those local copies. Then we can implement breaking changes in the `aws-cdk-lib` module and continue to release v1 as is. This lets us go back and add any critical updates to v1 modules, but obviously ends new features and minor bugfixes etc for v1 during the overlap period.

Reply:
I am concerned that duplicating the sources inside our repo and restructuring of our source repo will introduce risk and quite a lot of work into the critical path of v2.0. It will also require an awfully manual process to back port anything from 2.0 to 1.x and vice verse. 

I would also consider a git branch as a mechanism to maintain these two version lines temporarily. Perhaps we can even do that at an early stage in order to break the version number: branch off to a 2.x branch _today_ and have it continuously merge from master and build monocdk like today. Then, update the version number there to 2.x and create a separate pipeline that will release it.

Then, we can start applying changes and continue to merge from master until we officially release 2.x and perform the switch.

Reply:
Would something like this be feasible to keep it simple? Rather than trying to navigate a single branch driving two different release versions, we pick a date for the transition to a v2.0 release candidate, say the 'pre-release' date suggested above, and then on that date:

1. Create a v1.x branch of the repo. This will be the branch that the v1.x release(s) will be built from, and will only have security patches and important bugfixes backported into it. The v1.x releases of the CLI and Construct library are made off of the v1.x branch, and only the v1.x branch. This lets v1.x sticks around for a while for those that want to stay on it, but doesn't get any new features.

2. Mainline development shifts to a release-candidate process for v2.0. All new code targets the mainline branch. The CDK development team releases a new release candidate every N days/weeks, and when it looks stable enough for prime-time it does the official v2.0 release.

3. The reviewers for PRs on mainline would have to be cognizant of identifying changes that would need to be backported to the v1.x line, and there would have to be a process for that.

New v1.x releases are made as-needed, but their cadence should be drastically reduced as the vast majority of changes will be in mainline.  We don't break any existing apps that are out in the world because the v1.x line has a clean break from the v2.0 line.